libexec/pyenv-rehash

```diff
@@ -6,10 +6,44 @@ set -e
 
 SHIM_PATH="${PYENV_ROOT}/shims"
 PROTOTYPE_SHIM_PATH="${SHIM_PATH}/.pyenv-shim"
+LOCK_FILE_PATH=/var/lock/pyenv
+LOCK_FILE_FD=90
 
 # Create the shims directory if it doesn't already exist.
 mkdir -p "$SHIM_PATH"
 
+wait_for_lock() {
+  # flock(1) exists on Linux with util-linux and is the most reliable
+  if command -v flock &>/dev/null; then
+    eval "exec $LOCK_FILE_FD>\"$LOCK_FILE_PATH\""
+    flock --exclusive $LOCK_FILE_FD
+  # Fall back to shlock on BSD systems
+  elif command -v shlock &>/dev/null; then
+    # Try forever to acquire the lock
+    while ! shlock -f "$LOCK_FILE_PATH" -p $$; do
+      sleep 0.5
+    done
+  else
+    echo 'pyenv: cannot rehash: $PYENV_REHASH_PARALLEL_SAFE is set, but this system does not have flock or shlock.' >&2
+    exit 1
+  fi
+}
+
+release_lock() {
+  if command -v flock &>/dev/null; then
+    flock --unlock $LOCK_FILE_FD
+  elif command -v shlock &>/dev/null; then
+    rm -f $LOCK_FILE_PATH
+  fi
+}
+
+# When the PYENV_REHASH_PARALLEL_SAFE variable is set and flock(1) or shlock(1)
+# is available, this script can acquire an exclusive file lock to ensure only
+# one process is running pyenv-rehash.
+if [ -n "$PYENV_REHASH_PARALLEL_SAFE" ]; then
+  wait_for_lock
```

Comment:
There's some code for (pseudo) multi-process locking in `pyenv-rehash`. In general, it'd be better to avoid having code for similar purpose in multiple places. 

https://github.com/pyenv/pyenv/blob/a8e207f330509b12724454b1dd38dcc31193212f/libexec/pyenv-rehash#L13-L27

Please modify the code around the `noclobber` to,

1. Use `flock` or `shlock` if it is available on the platform (otherwise, fallback to `noclobber`)
1. Wait until the release of lock instead of immediate exit as error

For now it's immediately existing as error if there's another `pyenv-rehash` process is running. However, it must be some understandable behavior to wait until the release of lock since there're some of internal calls of `pyenv-rehash` within pyenv and they're actually NOT expecting `pyenv-rehash` to fail.

Reply:
Alright, I removed the environment variable to opt in to this behavior, and use `flock` or `shlock` when available. Unfortunately, without `flock` or `shlock`, there's no way to wait for the lock becoming available when we fall back to the `noclobber` variant. This limitation is fundamentally the reason for adding `flock` and `shlock` support.